### ISSUE TITLE
Change some configuration options into SX "alternative" rules

### ISSUE THREAD

**Issue by whut** [DotNetAnalyzers/StyleCopAnalyzers#1885]
This is just something to consider.

I believe that, like in StyleCop "classic", the rules should not be configurable, this way there is less confusion (and issues like #1847), less cases for maintaing stylecop.json for each project and more unification among different projects. In StyleCop "classic" not configurable rules was considered a feature.

Also it duplicates the concept of of SX.... "alternative" rules.

I checked all the current configuration options, here is the overview:
### usingDirectivesPlacement

SA1200 (UsingDirectivesMustBePlacedCorrectly) should behave only like in StyleCop "classic".
Putting usings inside namespace should be a separate SX... rule.
### systemUsingDirectivesFirst

SA1208 (SystemUsingDirectivesMustBePlacedBeforeOtherUsingDirectives) should be removed to follow current Visual Studio default, so that System directives are not first. This would be incompatible with StyleCop "classic", but I believe that "simple, non-configurable rules" approach, as in StyleCop "classis" is more important
### fileNamingConvention

SA1649 (FileNameMustMatchTypeName)  should behave only like in StyleCop "classic".
The "metadata" option should be a separate SX.... rule.
### newlineAtEndOfFile

SA1518 (UseLineEndingsCorrectlyAtEndOfFile) should only "require". This would be small incompability with StyleCop "classic", but in line with what most text editor and projects in other technologies do. Visual Studio by default puts this line at the end, so there should be not too many files to fix (checked on my current project, only few files had to be fixed).
The "omit" should be a separate SX.... rule.
### documentInterfaces,  documentExposedElements,  documentInternalElements,  documentPrivateElements  documentPrivateFields

SA1600 (ElementsMustBeDocumented) adds value by requiring internal members to be documented. For public members, the C# compiler will issue warning when XML documentation output is enabled, so there is no need to enable then this rule at all. For private members, I'm sure that no one document all of them.
### companyName,  copyrightText,  xmlHeader and variables

OK, as this is the only way to do it, same as in StyleCop "classic".
### elementOrder

OK as creating separate SX rules would be too cumbersone.

**Comment by sharwell** [comment]
I was originally of the same overall opinion. However, as things progressed it became clear that the "SX" rules approach did not offer a good long term solution.

1. With configuration, the expected behavior of rules is clear, regardless of which rules are enabled or disabled.
1. With configuration, if we ever have a release that has a bug in a rule that makes it misbehave, users can temporarily disable the rule without changing the behavior of other rules.
1. When rules are not flexible, it's hard to communicate to users that this project has broader appeal than StyleCop Classic due to its ability to check and correct alternative coding styles.

I came to realize that consistency within a project is more important than consistency across unrelated projects. This is not to say the latter is unimportant, as our default configuration is based on that principle.

Over time I'm hoping to remove all of the SX rules in favor of configuration. To start with I'm focusing on cases where enabling or disabling rules currently impacts the behavior of other rules. Before addressing other cases (eg SA1101 and SX1101) I'm waiting to see if Roslyn itself will provide a standard configuration format that we should share.

**Comment by yaakov-h** [comment]
> SA1200 (UsingDirectivesMustBePlacedCorrectly) should be changed to follow current Visual Studio default, so that it would NOT require System directives first.

I believe the current Visual Studio default is System first.

**Comment by sharwell** [comment]
The Visual Studio default *used* to be placing System using directives first. That was changed at some point, and an option was provided to restore the original behavior.

Changing the default behavior of SA1200 to not place System using directives first would force a large number of users to reformat their code. This is something we generally want to avoid. To date, the only change we made to the default behavior of SA1200 affects using directives starting with `global::System`, which is rare in practice.

**Comment by whut** [comment]
> Changing the default behavior of SA1200 to not place System using directives first would force a large number of users to reformat their code. 

This is exactly what Microsoft decided to do in Visual Studio 2012, they decided to change the default. [Here](https://connect.microsoft.com/VisualStudio/feedback/details/775702/organize-usings-no-longer-puts-system-references-first-bug-or-feature-change) is the reason why, wasn't that important, still they had no problem with changing it. As they did that, StyleCop should also have no problem with it (far less people use StyleCop than Visual Studio, so the impact is smaller), after all both with Visual Studio "Remove and sort using" and StyleCop codefix, it is just one click to update the source code.

**Comment by whut** [comment]
Allow me to disagree:

> 1. With configuration, the expected behavior of rules is clear, regardless of which rules are enabled or disabled.

I currently believe that a single rule that enforce exactly one style of coding is clearer, than rule that can enforce different approaches. The configuration is for "advanced users", people that want to change the already established and accepted so long ago by StyleCop community default. Each rule have some rationale behind it, there is a reason for example why "this" prefix is enforced, so adding configuration just obscures the rationale behind the rule and complicates documentation.

> 2. With configuration, if we ever have a release that has a bug in a rule that makes it misbehave, users can temporarily disable the rule without changing the behavior of other rules.

I'm not following this one, could you explain more? I currently understand is as "configuration allows me to disable a rule", which is not true, as rulesets are for it.

> 3. When rules are not flexible, it's hard to communicate to users that this project has broader appeal than StyleCop Classic due to its ability to check and correct alternative coding styles.

Not flexible rules is a StyleCop "classic" feature. But this flexibility is great addition to the "new" StyleCop, I'm really impressed seeing the SX rules and configuration options, that's something new and valuable.

I just wonder whenever configuration is the best approach to give that flexibility. I'm sure that configuration is needed for file headers and elements order (impressive!), but I'm trying to discuss if this complexity is needed for other rules. Please take a second look at my analysis above, doesn't the end result of moving more to SX rules would be simpler? 



BTW. I introduced StyleCop to few team and it was great experience for everyone to just avoid endless discussions about the coding style. Instead of arguing and forcing decisions on stubborn (there are always different points of view on everything in any team), we just had one question to answer: do we use StyleCop or not.

**Comment by sharwell** [comment]
> I'm not following this one, could you explain more? I currently understand is as "configuration allows me to disable a rule", which is not true, as rulesets are for it.

Certainly. The most obvious case is the way `UsingsCodeFixProvider` worked with `System` using directives. Currently we use the **systemUsingDirectivesFirst** configuration option, and SA1208, SA1210, and `UsingsCodeFixProvider` behave according to the specified value. Prior to this (specifically, prior to #1809), the behavior of `UsingsCodeFixProvider` actually changed based on whether or not SA1208 was disabled.

Now consider if we shipped 1.1.0 Beta 1 and there was a bug in SA1208 that resulted in AD0001 being reported. If users want to temporarily disable SA1208 so they can continue evaluating other parts of 1.1.0 Beta 1, we don't want this decision to start *changing* the behavior of other rules.

In summary, we want decisions about style to be independent of decisions about the order in which rules are enabled for a new project, such that the overall experience is consistent, flexible, and natural for new users.

**Comment by sharwell** [comment]
> BTW. I introduced StyleCop to few team and it was great experience for everyone to just avoid endless discussions about the coding style. Instead of arguing and forcing decisions on stubborn (there are always different points of view on everything in any team), we just had one question to answer: do we use StyleCop or not.

This is certainly a unique advantage of StyleCop Classic in some situations. One thing to remember for StyleCop Analyzers is when we evaluate potential new rules, one of our main discussion points is what a code fix would look like. By ensuring rules are crafted such that code fixes are straightforward and accurate, team members can easily adhere to the accepted coding styles without needing to understand the exact configuration that was chosen. We're working with the Roslyn analyzers team to improve this experience going forward, but nothing is set in stone yet.

**Comment by whut** [comment]
Thanks for explanations!

That is indeed bad that SA1210 depends on SA1208, but configuration option does not help. It only adds additional way of disabling SA1208 (like in the comment by @vweijsters in #1747). If there will be any issues with SA1208 requiring us to remove it (note that this alone is a very corner case, so I'm not sure if it should be optimized), regardless which way we choose to disable it (rule set or configuration option) it will still affect SA1210!

I think here the better approach would be to just remove the SA1208 rule altogether, as it is also outdated with regards to the Visual Studio default since 2012 release (I updated a bit my description of systemUsingDirectivesFirst option to reflect that).

Do you see any other issues like this with the other configuration options that are candidates for removal?

**Comment by sharwell** [comment]
Closing this since the project is moving in the other direction (removal of SX rules in favor of configuration).
